File: frontend/src/types/analysis.ts
// Mirrors the backend Pydantic schemas in app/schemas/analysis.py.
// Grows alongside the API contract as features are added.

export interface ImageInfo {
  filename: string;
  format: string;
  mode: string;
  width: number;
  height: number;
  megapixels: number;
  aspect_ratio: number;
  size_bytes: number;
}

export interface GpsCoordinates {
  latitude: number;
  longitude: number;
}

export interface ExifInfo {
  has_exif: boolean;
  make: string | null;
  model: string | null;
  lens: string | null;
  iso: number | null;
  aperture: string | null;
  shutter_speed: string | null;
  focal_length: string | null;
  date_taken: string | null;
  gps: GpsCoordinates | null;
}

export interface ColorSwatch {
  hex: string;
  rgb: number[];
  proportion: number;
}

export interface Dimensions {
  width: number;
  height: number;
  aspect_ratio: number;
}

export interface Histogram {
  bins: number;
  r: number[];
  g: number[];
  b: number[];
}

export interface DynamicRange {
  low: number;
  high: number;
  range: number;
  stops: number;
}

export type Orientation = "portrait" | "landscape" | "square";

export interface VisionInfo {
  brightness: number;
  contrast: number;
  sharpness: number;
  dominant_colors: ColorSwatch[];
  color_samples: number[][];
  histogram: Histogram;
  dynamic_range: DynamicRange;
  dimensions: Dimensions;
  orientation: Orientation;
}

export interface Point {
  x: number;
  y: number;
}

export interface BoundingBox {
  x0: number;
  y0: number;
  x1: number;
  y1: number;
}

export type SubjectSource = "detector" | "saliency" | "vlm";

/**
 * Maps a SubjectSource to a human-readable display label.
 *
 * The never check in the default branch makes TypeScript flag this function
 * as a compile error whenever a new SubjectSource variant is added without
 * a corresponding case — use this (or the same pattern) in any component
 * that switches on source, so regressions are caught at build time.
 */
export function subjectSourceLabel(source: SubjectSource): string {
  switch (source) {
    case "detector":
      return "Object Detector";
    case "saliency":
      return "Saliency Fallback";
    case "vlm":
      return "VLM";
    default: {
      const _exhaustive: never = source;
      return _exhaustive;
    }
  }
}

export interface RuleOfThirds {
  score: number;
  follows_rule: boolean;
  centroid: Point;
  nearest_power_point: Point;
  distance_to_power_point: number;
  source: SubjectSource;
}

export interface LineSegment {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
  angle: number;
}

export interface LeadingLines {
  has_leading_lines: boolean;
  line_count: number;
  dominant_angle: number | null;
  lines: LineSegment[];
}

export interface Horizon {
  horizon_detected: boolean;
  horizon_y: number | null;
  is_level: boolean;
  tilt_angle: number | null;
}

export interface Symmetry {
  vertical: number;
  horizontal: number;
  is_symmetric: boolean;
  dominant_axis: "vertical" | "horizontal";
}

export interface SubjectPosition {
  centroid: Point;
  region: string;
  offset_from_center: number;
  bbox: BoundingBox | null;
  label: string | null;
  confidence: number;
  has_mask: boolean;
  source: SubjectSource;
}

export interface EdgeRegions {
  top: number;
  bottom: number;
  left: number;
  right: number;
  center: number;
}

export interface EdgeDensity {
  edge_density: number;
  busyness: "minimal" | "moderate" | "busy";
  regions: EdgeRegions;
}

export interface NegativeSpace {
  negative_space_ratio: number;
  subject_excluded_ratio: number;
  has_significant_negative_space: boolean;
}

export interface CompositionInfo {
  rule_of_thirds: RuleOfThirds;
  leading_lines: LeadingLines;
  horizon: Horizon;
  symmetry: Symmetry;
  subject_position: SubjectPosition;
  edge_density: EdgeDensity;
  negative_space: NegativeSpace;
}

export interface AnalysisResponse {
  image: ImageInfo;
  exif: ExifInfo;
  vision: VisionInfo;
  composition: CompositionInfo;
}

// --- Phase 3: AI analysis --------------------------------------------------
// All content fields are optional: they come from a vision-language model and
// may be partially omitted, so the UI must render whatever is present.

export interface SceneInfo {
  summary: string | null;
  setting: string | null;
  tags: string[];
}

export interface SubjectInsight {
  primary: string | null;
  description: string | null;
}

export interface LightingInfo {
  summary: string | null;
  direction: string | null;
  quality: string | null;
  time_of_day: string | null;
}

export interface CameraSettings {
  aperture: string | null;
  shutter_speed: string | null;
  iso: string | null;
  focal_length: string | null;
  from_exif: boolean;
  reasoning: string | null;
}

export interface CompositionCritique {
  strengths: string[];
  improvements: string[];
  overall: string | null;
}

export interface SemanticLeadingLines {
  present: boolean | null;
  strength: number | null;
  description: string | null;
}

export interface SemanticScore {
  score: number | null;
  reasoning: string | null;
}

export interface SemanticComposition {
  leading_lines: SemanticLeadingLines | null;
  rule_of_thirds: SemanticScore | null;
  negative_space: SemanticScore | null;
}

export interface FujifilmRecipeSettings {
  grain: string | null;
  color_chrome_effect: string | null;
  white_balance: string | null;
  highlights: number | null;
  shadows: number | null;
  color: number | null;
  sharpness: number | null;
  noise_reduction: number | null;
}

export interface FujifilmRecipe {
  applicable: boolean | null;
  film_simulation: string | null;
  settings: FujifilmRecipeSettings | null;
  reasoning: string | null;
}

export interface AIAnalysis {
  available: boolean;
  reason: string | null;
  scene: SceneInfo | null;
  subject: SubjectInsight | null;
  lighting: LightingInfo | null;
  camera_settings: CameraSettings | null;
  composition_critique: CompositionCritique | null;
  recreation_guide: string[];
  semantic_composition: SemanticComposition | null;
  fujifilm_recipe: FujifilmRecipe | null;
}

export interface AIAnalysisResponse {
  ai: AIAnalysis;
}

// --- Color grading -----------------------------------------------------

export interface GradingAdjustments {
  exposure: number;
  contrast: number;
  highlights: number;
  shadows: number;
  whites: number;
  blacks: number;
  temperature: number;
  tint: number;
  saturation: number;
  vibrance: number;
  sharpness: number;
}

export type ColorGradeStyle =
  | "natural"
  | "cinematic"
  | "moody"
  | "airy"
  | "contrasty"
  | "warm"
  | "cool";

export interface ColorGradeResponse {
  available: boolean;
  adjustments: GradingAdjustments;
  reasoning: string | null;
  style: ColorGradeStyle | null;
  reason: string | null;
}

File: backend/app/schemas/analysis.py
"""API contract for the analysis endpoint.

`AnalysisResponse` is intentionally small right now. Each feature we add
(EXIF, composition, lighting, AI critique...) appends a new optional field
here, so the contract grows without breaking existing clients.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class ImageInfo(BaseModel):
    filename: str
    format: str
    mode: str
    width: int
    height: int
    megapixels: float
    aspect_ratio: float
    size_bytes: int


class GpsCoordinates(BaseModel):
    latitude: float
    longitude: float


class ExifInfo(BaseModel):
    """Camera metadata read from the file. Every field is optional because
    many images (screenshots, exports, stripped photos) carry no EXIF."""

    has_exif: bool = Field(..., description="True if any EXIF field was found.")
    make: str | None = None
    model: str | None = None
    lens: str | None = None
    iso: int | None = None
    aperture: str | None = None
    shutter_speed: str | None = None
    focal_length: str | None = None
    date_taken: str | None = None
    gps: GpsCoordinates | None = None


class ColorSwatch(BaseModel):
    hex: str
    rgb: list[int]
    proportion: float


class Dimensions(BaseModel):
    width: int
    height: int
    aspect_ratio: float


class Histogram(BaseModel):
    bins: int
    r: list[int]
    g: list[int]
    b: list[int]


class DynamicRange(BaseModel):
    low: float
    high: float
    range: float
    stops: float


class VisionInfo(BaseModel):
    """Objective image-quality metrics from the OpenCV pipeline (no AI)."""

    brightness: float
    contrast: float
    sharpness: float
    dominant_colors: list[ColorSwatch]
    color_samples: list[list[int]] = Field(
        default_factory=list,
        description="Raw [r, g, b] pixel scatter for the 3D color-space point cloud.",
    )
    histogram: Histogram
    dynamic_range: DynamicRange
    dimensions: Dimensions
    orientation: str


class Point(BaseModel):
    x: float
    y: float


class BoundingBox(BaseModel):
    x0: float
    y0: float
    x1: float
    y1: float


class RuleOfThirds(BaseModel):
    score: float
    follows_rule: bool
    centroid: Point
    nearest_power_point: Point
    distance_to_power_point: float
    source: str = "saliency"


class LineSegment(BaseModel):
    x1: int
    y1: int
    x2: int
    y2: int
    angle: float


class LeadingLines(BaseModel):
    has_leading_lines: bool
    line_count: int
    dominant_angle: float | None
    lines: list[LineSegment]


class Horizon(BaseModel):
    horizon_detected: bool
    horizon_y: float | None
    is_level: bool
    tilt_angle: float | None


class Symmetry(BaseModel):
    vertical: float
    horizontal: float
    is_symmetric: bool
    dominant_axis: str


class SubjectPosition(BaseModel):
    centroid: Point
    region: str
    offset_from_center: float
    bbox: BoundingBox | None = None
    label: str | None = None
    confidence: float = 0.0
    has_mask: bool = False
    source: str = "saliency"


class EdgeRegions(BaseModel):
    top: float
    bottom: float
    left: float
    right: float
    center: float


class EdgeDensity(BaseModel):
    edge_density: float
    busyness: str
    regions: EdgeRegions


class NegativeSpace(BaseModel):
    negative_space_ratio: float
    subject_excluded_ratio: float
    has_significant_negative_space: bool


class CompositionInfo(BaseModel):
    """Geometric/structural composition metrics from OpenCV (no AI)."""

    rule_of_thirds: RuleOfThirds
    leading_lines: LeadingLines
    horizon: Horizon
    symmetry: Symmetry
    subject_position: SubjectPosition
    edge_density: EdgeDensity
    negative_space: NegativeSpace


class AnalysisResponse(BaseModel):
    image: ImageInfo = Field(..., description="Basic metadata about the upload.")
    exif: ExifInfo = Field(..., description="Camera metadata extracted from EXIF.")
    vision: VisionInfo = Field(..., description="Objective CV image-quality metrics.")
    composition: CompositionInfo = Field(
        ..., description="Geometric composition metrics."
    )


# --- Phase 3: AI analysis --------------------------------------------------
# Every field on the sub-models is optional: the response is produced by a
# vision-language model, so a partial/omitted field must degrade gracefully
# rather than fail validation.


class SceneInfo(BaseModel):
    summary: str | None = None
    setting: str | None = None
    tags: list[str] = Field(default_factory=list)


class SubjectInsight(BaseModel):
    primary: str | None = None
    description: str | None = None


class LightingInfo(BaseModel):
    summary: str | None = None
    direction: str | None = None
    quality: str | None = None
    time_of_day: str | None = None


class CameraSettings(BaseModel):
    aperture: str | None = None
    shutter_speed: str | None = None
    iso: str | None = None
    focal_length: str | None = None
    from_exif: bool = False
    reasoning: str | None = None


class CompositionCritique(BaseModel):
    strengths: list[str] = Field(default_factory=list)
    improvements: list[str] = Field(default_factory=list)
    overall: str | None = None


class SemanticLeadingLines(BaseModel):
    present: bool | None = None
    strength: float | None = None
    description: str | None = None


class SemanticScore(BaseModel):
    score: float | None = None
    reasoning: str | None = None


class SemanticComposition(BaseModel):
    """The AI's meaning-aware read of composition. Intentionally replaces the
    geometric CV scores for these three dimensions."""

    leading_lines: SemanticLeadingLines | None = None
    rule_of_thirds: SemanticScore | None = None
    negative_space: SemanticScore | None = None


class FujifilmRecipeSettings(BaseModel):
    grain: str | None = None
    color_chrome_effect: str | None = None
    white_balance: str | None = None
    highlights: float | None = None
    shadows: float | None = None
    color: float | None = None
    sharpness: float | None = None
    noise_reduction: float | None = None


class FujifilmRecipe(BaseModel):
    """A Fujifilm film-simulation recipe recommendation (Phase 4).

    ``applicable`` is False when the shot was not taken on a Fujifilm body."""

    applicable: bool | None = None
    film_simulation: str | None = None
    settings: FujifilmRecipeSettings | None = None
    reasoning: str | None = None


class AIAnalysis(BaseModel):
    """Vision-language interpretation of the photo (Phase 3).

    ``available`` is False when AI analysis is unconfigured or failed; in that
    case ``reason`` explains why and the content fields are omitted.
    """

    available: bool
    reason: str | None = None
    scene: SceneInfo | None = None
    subject: SubjectInsight | None = None
    lighting: LightingInfo | None = None
    camera_settings: CameraSettings | None = None
    composition_critique: CompositionCritique | None = None
    recreation_guide: list[str] = Field(default_factory=list)
    semantic_composition: SemanticComposition | None = None
    fujifilm_recipe: FujifilmRecipe | None = None


class AIAnalysisResponse(BaseModel):
    ai: AIAnalysis = Field(..., description="AI interpretation of the photo.")


# --- Color grading -----------------------------------------------------
# A single Claude call suggesting slider-ready adjustments. Mirrors
# AIAnalysis: every field degrades gracefully, and `available=False` on
# any failure rather than raising.


class GradingAdjustments(BaseModel):
    exposure: float = 0.0
    contrast: float = 0.0
    highlights: float = 0.0
    shadows: float = 0.0
    whites: float = 0.0
    blacks: float = 0.0
    temperature: float = 0.0
    tint: float = 0.0
    saturation: float = 0.0
    vibrance: float = 0.0
    sharpness: float = 0.0


class ColorGradeResponse(BaseModel):
    available: bool
    adjustments: GradingAdjustments = Field(default_factory=GradingAdjustments)
    reasoning: str | None = None
    style: str | None = None
    reason: str | None = None
